### DISCUSSION TITLE
Run step on (if branch/tag is)

### DISCUSSION THREAD

**Discussion by bhowell2** [community/community#25216] [Actions]
<p>Currently, I do not believe there is a way to run a step (or even a job) if a branch matches a certain pattern. It is only possible to run an entire workflow on a master/tagged branch, but this requires duplicating the workflow just to add another step for releasing/publishing. I could also create an action for the build/test steps, but both creating an action and duplicating the code seem superfulous and more likely to cause errors since someone could forget that both files need to be edited.</p>
<p>A common example of this would be:</p>
<ul>
<li>
<p>Checkout</p>
</li>
<li>
<p>Run tests/whatever</p>
</li>
<li>
<p>IF on master/tagged branch THEN release/publish code.</p>
</li>
</ul>
<p>Am i missing something and there is an easy way to do this?</p>
<p>I know i can do:</p>
<pre><code>if: {{ github.ref == 'ref/head/master' }}
</code></pre>
<p>but that only works when you know the branch/tag exactly and something like (with pattern):</p>
<pre><code>if: {{ github.ref == 'ref/head/v*' }}
</code></pre>
<p>does not work.</p>
<p>Right now the best approach I can think of is to create some type of conditional-step-action that will allow you to input your if condition and then run the step if the branch/tag condition matches, but exit without failing if it does not (so as not to fail the job).</p>

**Comment by BradenM** [comment, marked as answer]
<p>Hi <a class="mention" href="/bhowell2">@bhowell2</a>,</p>
<p>First, you do not need to use conditional syntax with the “if” statement. See the following excerpt from: "<a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/contexts-and-expression-syntax-for-github-actions#about-context-and-expressions" rel="nofollow noopener">About context and expressions"</a></p>
<p>When you use expressions in an if conditional, you do not need to use the expression syntax ({{ }}) because GitHub automatically evaluates the if conditional as an expression.</p>
<p>As for pattern matching with the if statement, I suggest using the <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/contexts-and-expression-syntax-for-github-actions#contains" rel="nofollow noopener">contains</a> function. Here’s an example:</p>
<pre><code>if: contains(github.ref, "release")
steps:
 - run: echo "I only run if the branch has release in its name!"
</code></pre>

**Reply by karlhorky** [reply]
@BradenM thanks for this!

Tried to achieve the opposite and it seems like the opposite (NOT this branch) can also be achieved by using the `!` before the `contains` function and quoting the whole expression:

```yaml
if: "!contains(github.ref, 'main')"
```

**Reply by gamesguru** [reply]
You can combine statements too, e.g.

```yaml
if: contains(github.ref, 'release') || github.ref == 'refs/heads/master'
```

**Reply by sankansm23** [reply]
But does it work for such condition ? For branch with names like release/1.0.15
if: contains(github.ref, "release/*")
steps:
 - run: echo "I only run if the branch has release in its name!"

**Comment by bhowell2** [comment]
<p><span class="mention">@bradenmthanks</span> for the response (and noted on the if expression!). Contains does not do pattern matching, but it could definitely work. Using contains would require people to name their release tags something that is unlikely to occur elsewhere (luckily tags are generally for some type of releases anyway). As you put it “release” is a pretty good indicator that the tag is for a release. Many currently use something more of the form “v*” or maybe just a semantic versioned number (e.g. 1.0.1), but probably wouldn’t be a huge deal for people to adjust this by adding “release”, “milestone” or whatever in the tag. This would be much less of a problem anyway if workflows could be reused/referenced which would be preferred to using the IF condition at all.</p>

**Comment by rowi1de** [comment]
<p>You can use bash substitutions:</p>
<p>e.g. </p>
<pre><code>name: Extract Branch Name
run: echo "::set-env name=BRANCH::$(echo ${GITHUB_REF##*/})"
</code></pre>
<p>would strip refs/heads and give you “master” </p>
<p>You should be able to use also RegExp here.</p>
<p>Also see <a href="https://github.community/t/implementing-dry-run-logic-if-and-env/17834">https://github.community/t5/GitHub-Actions/Implementing-Dry-Run-Logic-If-and-env/td-p/49882</a></p>

**Comment by jef** [comment]
<p>For those looking at this article today and onward, please use the <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable" rel="noopener nofollow ugc">new declaration</a> for setting environment variables:</p>
<pre><code class="lang-yaml">- name: Extract Branch Name
  run: echo "BRANCH=$(echo ${GITHUB_REF##*/})" &gt;&gt; $GITHUB_ENV
</code></pre>

**Comment by fkirc** [comment]
<p>You could fork <a href="https://github.com/fkirc/skip-duplicate-actions" rel="noopener nofollow ugc">https://github.com/fkirc/skip-duplicate-actions</a> to get branch-specific skipping. With such a fork, it should be doable with less than 5 lines of code.</p>

**Comment by TomasVotruba** [comment]
<p>This is epic, thank you for sharing the $GITHUB_ENV trick!</p>

**Comment by kamranayub** [comment]
<p>Just FYI for anyone else using this as a reference, the values should be <code>refs/heads/master</code> not <code>ref/head</code> :slight_smile:</p>

**Comment by astrochun** [comment]
<p>Note that this should be single quotes around release. Otherwise you will get an invalid workflow warning. For example:</p>
<pre><code class="lang-auto">The workflow is not valid. .github/workflows/docker-action-check.yml (Line: 24, Col: 9): Unexpected symbol: '"refs/tags"'
</code></pre>

**Reply by patrickelectric** [reply]
> Note that this should be single quotes around release. Otherwise you will get an invalid workflow warning. For example:
> 
> ```
> The workflow is not valid. .github/workflows/docker-action-check.yml (Line: 24, Col: 9): Unexpected symbol: '"refs/tags"'
> ```

I hope that there is a good explanation about that from github..
